### PR TITLE
Fix for issue #13754

### LINKED ISSUE #13754: --config-scope seems to be ignored
Starting with commit 43b0356e14a8227f33040302cf948d3c98508d59, configuration scopes seem to be ignored.  For example, with one commit before (dc36b4737d16ea4aa49daa14fb782b5ea85b60e3):

```console
$ cd $spack_root
$ git checkout dc36b4737d16ea4aa49daa14fb782b5ea85b60e3
$ ./bin/spack --config-scope=<some path> config blame repos
---                                            repos:
<some path>/repos.yaml:2                       - $nevada_toolset2_root/var/nevada/spack/repo
<spack root>/etc/spack/defaults/repos.yaml:14  - $spack/var/spack/repos/builtin
```

but, after 43b0356e14a8227f33040302cf948d3c98508d59 I get

```console
$ cd $spack_root
$ git checkout develop
$ ./bin/spack --config-scope=<some path> config blame repos
---                                            repos:
<spack root>/etc/spack/defaults/repos.yaml:14  - $spack/var/spack/repos/builtin
```

Checking out `lib/spack/spack/relocate.py` from dc36b4737d16ea4aa49daa14fb782b5ea85b60e3 seems to resolve the issue:

```console
$ cd $spack_root
$ git co dc36b4737d16ea4aa49daa14fb782b5ea85b60e3 -- lib/spack/spack/relocate.py
$ ./bin/spack --config-scope=<some path> config blame repos
---                                            repos:
<some path>/repos.yaml:2                       - $nevada_toolset2_root/var/nevada/spack/repo
<spack root>/etc/spack/defaults/repos.yaml:14  - $spack/var/spack/repos/builtin
```

One of the changes in 43b0356e14a8227f33040302cf948d3c98508d59 is to the function signature of `file_is_relocatable` from

```python
def file_is_relocatable(file):
    ...
```

To

```python
def file_is_relocatable(
    file, paths_to_relocate=[spack.store.layout.root, spack.paths.prefix]):
```

`paths_to_relocate` is then used in the function.  Simply making the following change fixes the config scope issue for me:

```python
def file_is_relocatable(file):
    paths_to_relocate = [spack.store.layout.root, spack.paths.prefix]
```

That said, this makes no sense to me.  The function `file_is_relocatable` is not called when `spack config blame ...` is issued (tested by dropping `assert 0` as the first line of `file_is_relocatable`).  Nevertheless, the change does fix my issue.

A grep of the spack source shows:

```console
$ cd $spack_root
$ grep -irl  --include=*.py file_is_relocatable .
./lib/spack/spack/test/relocate.py
./lib/spack/spack/binary_distribution.py
./lib/spack/spack/relocate.py
```

and none of the calls use the `paths_to_relocate` argument, so the change above should be fine, and it might be best to leave mutable arguments from function signature, if possible.

@becker33, since he was the author of 43b0356e

### PR DESCRIPTION
**Pull request by tjfulle**
fixes #13754

### CODE REVIEW

**Comment by tjfulle** [comment]
@becker33 (since he is mentioned in #13754)

**Comment by chissg** [comment]
@gartung

**Comment by gartung** [comment]
@becker33

File: lib/spack/spack/relocate.py [hunk 1]
```diff
@@ -651,8 +651,7 @@ def is_relocatable(spec):
     return True
 
 
-def file_is_relocatable(
-        file, paths_to_relocate=[spack.store.layout.root, spack.paths.prefix]):
```
Comment (alalazo) [inline lib/spack/spack/relocate.py]: What happens if you turn this into a tuple instead? Not having lists as defaults for function arguments is a Python's gotcha, so I am wondering if this is the real issue here.

File: lib/spack/spack/relocate.py [hunk 2]
```diff
@@ -651,8 +651,7 @@ def is_relocatable(spec):
...
+def file_is_relocatable(file):
```
Comment (becker33) [inline lib/spack/spack/relocate.py]: This change will break binary relocation.

If the issue is lists as default arguments, we should change this to `def file_is_relocatable(file, paths_to_relocate=None)` and insert the defaults below if we see `paths_to_relocate is None`

**Comment by scheibelp** [comment]
Using a tuple doesn't appear to resolve the issue, but using `None` (and then initializing in the function body) as suggested by @becker33 does appear to also resolve the issue (so it does appear to be a problem of the form suggested by @alalazo).

File: lib/spack/spack/relocate.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/relocate.py]: Just checked. Apparently that's not the issue here.

**Comment by alalazo** [comment]
@scheibelp I'm trying to understand what's going on. Tuples are immutable so it should be fine to use them as defaults.

**Comment by tjfulle** [comment]
The fact that these changes make a difference at all is strange to me since this code path does not seem to be even exercised in the example shown in #13754.

**Comment by becker33** [comment]
Yeah I'm quite confused by this as well. I've confirmed the bug and that my version of the fix resolves the bug without losing the functionality, but I have no idea why this is happening.

**Comment by alalazo** [comment]
Okay, this is tricky. The issue seems to be due to `spack.store.<something>` being in the default values. This triggers a call to `spack.config` at module definition time, i.e. before we even enter `main` and surely before the command line is processed. This information seems to be cached and re-used later. Makes sense?

**Comment by tjfulle** [comment]
Perfect sense...  All the more reason to adopt:

```python
def file_is_relocatable(file, paths_to_relocate=None):
    if paths_to_relocate is None:
        paths_to_relocate = ...
```

**Comment by alalazo** [comment]
@tjfulle Or the more idiomatic equivalent:
```python
def file_is_relocatable(file, paths_to_relocate=None):
    paths_to_relocate = paths_to_relocate or default_value
```

File: lib/spack/spack/relocate.py (hunk 2 above)
Comment (tjfulle) [reply lib/spack/spack/relocate.py]: @becker33, see latest change.

**Pull request merged**